pkg/apis/configuration/v1alpha1/kongcustomentity_types.go

```diff
@@ -0,0 +1,76 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// KongCustomEntityList is a list of kong custom entities.
+type KongCustomEntityList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KongCustomEntity `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kongcustomentities,shortName=kce,categories=kong-ingress-controller
+
+// KongCustomEntity represents a custom entity in Kong.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Type is the type of this custom entity.
+	// Should be same as the `Name` of a KongCustomEntityDefinition.
+	Type string `json:"type"`
+	// Fields is the list of fields in the entity.
+	Fields apiextensionsv1.JSON `json:"fields"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,path=kongcustomentitydefinitions,shortName=kced,categories=kong-ingress-controller
+
+// KongCustomEntityDefinition represents definition of a custom entity type in Kong.
+type KongCustomEntityDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Name is the type name of the entity.
+	Name string `json:"name"`
+	// AdminAPIName is the name used in admin API paths to CRUD this type of entity.
+	// If AdminAPIName is empty, it is seen as same as `Name`.
+	// For example: Name = "jwt_credentials" and AdminAPIName = "jwts", then we call `/jwts` or `/jwts/<id>` for CRUD.
+	AdminAPIName string `json:"adminAPIName,omitempty"`
```

Comment:
What about the nested admin API name? Do I understand correctly it won't be really needed in decK because we may use the admin API name only and define the relation in the request?

Reply:
For usage of decK or declarative configuration, I think nested API name is not required. But I think it will do at least no harm to add this field. So I think this should be added.

Reply:
Hmm, I think that it will be better to not add it until we find out it's required for something. If we discover that it's needed somewhere in decK to properly sync, we will just add it here.

Reply:
This is the case for declarative. Are we sure for deck? Without the implementation available it's not really possible to say, but the closest equivalent (credentials) [do use the nested path](https://github.com/Kong/go-kong/blob/a8312cc37af248060530fa4197a4360b13a6d37d/kong/credentials_service.go#L28-L54).

Using the root path requires knowing the attached entity's ID. You can technically handle that by creating core entities first and using a lookup, and AFAIK this is how routes work (deck always creates the services first and uses the root `/routes` endpoint), but it's hard to say which will be easier to implement in practice. Using the parent name (and thus the nested endpoint) feels easier since you don't need any lookups, just the name from the manifest.

Either way, it is easy to add or remove the field before release.

Reply:
Added `AdminAPINestedName` here.